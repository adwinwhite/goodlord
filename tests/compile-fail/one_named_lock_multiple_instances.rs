use std::thread;
use std::sync::Arc;

use goodlord::LockAfter;
use goodlord::impl_lock_order;
use goodlord::LockToken;
use goodlord::GoodLock;
use goodlord::Unlocked;


struct A;

impl_lock_order!(Unlocked => A);

fn main() {
    let mut unlocked_token = LockToken::new();
    let lock_a1: Arc<GoodLock<A, _>> = GoodLock::new(1_i32).into();
    let lock_a2: Arc<GoodLock<A, _>> = GoodLock::new(2_i32).into();
    let handle = {
        let lock_a1 = lock_a1.clone();
        let lock_a2 = lock_a2.clone();
        thread::spawn(move || {
            let mut unlocked_token = LockToken::new();
            let (a1, a1_token) = lock_a1.lock(&mut unlocked_token);
            // We cannot use a1_token to lock lock_a2 due to we can't have `A => A`.
            let (a2, _) = lock_a2.lock(&mut a1_token);
            println!("{}, {}", a1, a2);
        })
    };
    let (a2, _) = lock_a2.lock(&mut unlocked_token);
    // We cannot use a token for multiple times in the same scope.
    let (a1, _) = lock_a1.lock(&mut unlocked_token);
    handle.join();
    println!("{}, {}", a2, a1);
}
