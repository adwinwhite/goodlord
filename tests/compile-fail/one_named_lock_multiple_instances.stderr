error[E0308]: mismatched types
  --> tests/compile-fail/one_named_lock_multiple_instances.rs:26:40
   |
26 |             let (a2, _) = lock_a2.lock(&mut a1_token);
   |                                   ---- ^^^^^^^^^^^^^ expected `&mut LockToken<Unlocked>`, found `&mut LockToken<A>`
   |                                   |
   |                                   arguments to this method are incorrect
   |
   = note: expected mutable reference `&mut LockToken<Unlocked>`
              found mutable reference `&mut LockToken<A>`
note: method defined here
  --> src/lib.rs
   |
   |     pub fn lock<'t, 's, PrevId>(
   |            ^^^^
